services/twofa.go

```diff
@@ -0,0 +1,283 @@
+	}
+
+	return &TwoFAService{
+		issuer:           issuer,
+		qrSize:           qrSize,
+		ds:               ds,
+		keyServiceClient: client,
+		isKeyService:     isKeyService,
+	}
+}
+
+// DisableTwoFA disables two-factor authentication for an account
+func (t *TwoFAService) DisableTwoFA(accountID uuid.UUID) error {
+	if err := t.ds.SetTOTPSetting(accountID, false); err != nil {
+		return err
+	}
+
+	if err := t.DeleteTOTPKey(accountID); err != nil {
+		return err
+	}
+
+	if err := t.ds.SetRecoveryKey(accountID, nil); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ProcessAuthRequest verifies either TOTP code or recovery key for an account
+func (t *TwoFAService) ProcessAuthRequest(loginState *datastore.InterimPasswordState, req *TwoFAAuthRequest) error {
+	// Verify either TOTP code or recovery key
+	if req.TOTPCode != nil {
+		// Verify TOTP code
+		if err := t.ValidateTOTPCode(*loginState.AccountID, *req.TOTPCode); err != nil {
+			return err
+		}
+	} else if req.RecoveryKey != nil {
+		// Verify recovery key
+		if err := t.ds.CheckRecoveryKey(*loginState.AccountID, *req.RecoveryKey); err != nil {
+			return err
+		}
+		if err := t.DisableTwoFA(*loginState.AccountID); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// GenerateAndStoreTOTPKey creates and stores a new TOTP key for an account
+func (t *TwoFAService) GenerateAndStoreTOTPKey(accountID uuid.UUID, email string) (*otp.Key, error) {
+	if t.keyServiceClient != nil {
+		return t.makeKeyServiceTOTPGenerateRequest(accountID, email)
+	}
+
+	// Otherwise, generate and store the key locally
+	opts := totp.GenerateOpts{
+		Issuer:      t.issuer,
+		AccountName: email,
+	}
+
+	key, err := totp.Generate(opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate TOTP key: %w", err)
+	}
+
+	if err := t.ds.StoreTOTPKey(accountID, key); err != nil {
+		return nil, fmt.Errorf("failed to store TOTP key: %w", err)
+	}
+
+	return key, nil
+}
+
+// makeKeyServiceTOTPGenerateRequest sends a request to the key service to generate and store a TOTP key
+func (t *TwoFAService) makeKeyServiceTOTPGenerateRequest(accountID uuid.UUID, email string) (*otp.Key, error) {
+	type totpGenerateRequest struct {
+		AccountID uuid.UUID `json:"accountId"`
+		Email     string    `json:"email"`
+	}
+
+	type totpGenerateResponse struct {
+		URL string `json:"url"`
+	}
+
+	reqBody := totpGenerateRequest{
+		AccountID: accountID,
+		Email:     email,
+	}
+
+	var response totpGenerateResponse
+	if err := t.keyServiceClient.MakeRequest(http.MethodPost, "/v2/server_keys/totp", reqBody, &response); err != nil {
+		return nil, err
+	}
+
+	// Parse the URL to create an OTP key
+	key, err := otp.NewKeyFromURL(response.URL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse TOTP key URL: %w", err)
+	}
+
+	return key, nil
+}
+
+// ValidateTOTPCode checks if the provided code is valid for the specified account
+func (t *TwoFAService) ValidateTOTPCode(accountID uuid.UUID, code string) error {
```

Comment:
[RFC 6238 notes](https://datatracker.ietf.org/doc/html/rfc6238#section-4) that one-time codes should only be usable once within an application:
> Note that a prover may send the same OTP inside a given time-step window multiple times to a verifier.  The verifier MUST NOT accept the second attempt of the OTP after the successful validation has been issued for the first OTP, which ensures one-time only use of an OTP.

This suggests that we should have a cache of codes that have been used successfully by a given accountID. This cache can set a TTL of 2 min on the entries.